irohad/torii/impl/command_service_impl.cpp

```diff
@@ -174,62 +174,34 @@ namespace iroha {
-          // another for statuses pushing. That can allow to move a part of code
-          // to a separate method for simplification
-          std::accumulate(
-              txs.begin(),
-              txs.end(),
-              std::make_pair<bool, bool>(false, false),
-              [this, &status_issuer](std::pair<bool, bool> lookup_result,
-                                     const auto &tx) {
-                const auto &tx_hash = tx->hash();
-                if (auto found = cache_->findItem(tx_hash)) {
-                  iroha::visit_in_place(
-                      (*found)->get(),
-                      [this, &found, &lookup_result, &status_issuer](
-                          const shared_model::interface::MstPendingResponse &) {
-                        this->pushStatus(status_issuer, *found);
-                        lookup_result.second = true;
-                      },
-                      [this, &tx_hash, &status_issuer](
-                          const shared_model::interface::NotReceivedTxResponse
-                              &) {
-                        // This branch covers an impossible case (this cache
-                        // cannot contain NotReceivedTxResponse, because the tx
-                        // has reached processBatch method, which means that
-                        // the tx already has StatelessValid status).
-                        // That is why we are not updating its status inside
-                        // internal cache, but still pushing to status bus.
-                        this->pushStatus(
-                            status_issuer,
-                            status_factory_->makeStatelessValid(tx_hash));
-                      },
-                      [this, &found, &status_issuer](const auto &status) {
-                        this->pushStatus(status_issuer, *found);
-                      });
-                  lookup_result.first = true;
-                }
-                return lookup_result;
-              });
-
-      if (cache_has_at_least_one_tx and not batch_has_mst_pending_tx) {
-        // If a non-persistent cache says that a transaction has pending status
-        // that means we have to check persistent cache too.
-        // Non-persistent cache might be overflowed and mst replay become
-        // possible without checking persistent cache.
-
-        // If there are no pending statuses and the transaction is found in
-        // non-persistent cache, then it is considered as a replay and prevented
-        // from further propagation.
+      bool has_final_status{false};
+
+      for (auto tx : txs) {
+        const auto &tx_hash = tx->hash();
+        if (auto found = cache_->findItem(tx_hash)) {
+          log_->debug("Found in cache: {}", **found);
+          has_final_status = iroha::visit_in_place(
+              (*found)->get(),
+              [](const shared_model::interface::CommittedTxResponse &) {
```

Comment:
Another option is to reuse `FinalStatusValue`:
https://github.com/hyperledger/iroha/blob/204a61960b772e85f6a38fa47ddefdec5bf99d61/irohad/torii/impl/command_service_impl.cpp#L152-L159